src/App.jsx

```diff
@@ -14,45 +16,105 @@ export const goodsFromServer = [
-
-      <button
-        type="button"
-        className="button is-warning is-light"
-      >
-        Reverse
-      </button>
-
-      <button
-        type="button"
-        className="button is-danger is-light"
-      >
-        Reset
-      </button>
-    </div>
+export const App = () => {
+  const [goods, setGoods] = useState(initialGoods);
+  const [buttonStates, setButtonStates] = useState({
+    sortAlphabetically: true,
+    sortByLength: true,
+    reverse: true,
+    reset: true,
+  });
+  const [ascending, setAscending] = useState(true);
+  const [originalOrder] = useState([...initialGoods]); // Store the original order during component initialization
+
+  const handleButtonClick = (buttonName) => {
+    const newButtonStates = { ...buttonStates };
+
+    Object.keys(newButtonStates).forEach((key) => {
+      newButtonStates[key] = key === buttonName;
+    });
+    setButtonStates(newButtonStates);
+  };
+
+  const sortAlphabetically = () => {
+    const sortedGoods = [...goods].sort();
+
+    setGoods(sortedGoods);
+    setAscending(true);
+    handleButtonClick('sortAlphabetically');
+  };
+
+  const sortByLengthStable = () => {
+    const sortedGoods = [...goods].sort((a, b) => {
+      if (a.length !== b.length) {
+        return a.length - b.length;
+      }
+
+      return goods.indexOf(a) - goods.indexOf(b);
+    });
+
+    setGoods(sortedGoods);
+    setAscending(true);
+    handleButtonClick('sortByLength');
+  };
 
-    <ul>
-      <li data-cy="Good">Dumplings</li>
-      <li data-cy="Good">Carrot</li>
-      <li data-cy="Good">Eggs</li>
-      <li data-cy="Good">Ice cream</li>
-      <li data-cy="Good">Apple</li>
-      <li data-cy="Good">...</li>
-    </ul>
-  </div>
-);
+  const reverseOrder = () => {
+    const reversedGoods = [...goods].reverse();
+
+    setGoods(reversedGoods);
+    setAscending(!ascending);
+    handleButtonClick('reverse');
+  };
+
+  const resetOrder = () => {
+    setGoods(originalOrder);
+    setAscending(true);
+    handleButtonClick('reset');
+  };
+
+  return (
+    <div className="section content">
+      <div className="buttons">
+        <button
+          type="button"
+          className={`button is-info ${buttonStates.sortAlphabetically ? '' : 'is-light'}`}
+          onClick={sortAlphabetically}
+        >
+          Sort alphabetically
+        </button>
+
+        <button
+          type="button"
+          className={`button is-success ${buttonStates.sortByLength ? '' : 'is-light'}`}
+          onClick={sortByLengthStable}
+        >
+          Sort by length
+        </button>
+
+        <button
+          type="button"
+          className={`button is-warning ${buttonStates.reverse ? '' : 'is-light'}`}
+          onClick={reverseOrder}
+        >
+          Reverse
+        </button>
+
+        <button
+          type="button"
+          className={`button is-danger ${buttonStates.reset ? '' : 'is-light'}`}
+          onClick={resetOrder}
+        >
+          Reset
```

Comment:
There is no need to render the reverse button in all cases. Consider rendering it if sort or order is applied